nav2_map_server/src/vo_server/vector_object_shapes.cpp

```diff
@@ -0,0 +1,579 @@
+// Copyright (c) 2023 Samsung R&D Institute Russia
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_map_server/vector_object_shapes.hpp"
+
+#include <uuid/uuid.h>
+#include <math.h>
+#include <cmath>
+#include <limits>
+#include <stdexcept>
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+
+#include "nav2_util/occ_grid_utils.hpp"
+#include "nav2_util/occ_grid_values.hpp"
+#include "nav2_util/raytrace_line_2d.hpp"
+#include "nav2_util/robot_utils.hpp"
+
+namespace nav2_map_server
+{
+
+// ---------- Shape ----------
+
+Shape::Shape(const nav2_util::LifecycleNode::WeakPtr & node)
+: type_(UNKNOWN), node_(node)
+{}
+
+Shape::~Shape()
+{}
+
+ShapeType Shape::getType()
+{
+  return type_;
+}
+
+bool Shape::getShapeUUID(const std::string & shape_name, unsigned char * out_uuid)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    // Try to get shape UUID from ROS-parameters
+    std::string uuid_str = getROSParameter(
```

Comment:
What if this is a shape set by a service call?

Reply:
During the creation of new Polygon/Circle shape, parameters are being passed as an input argument from a service call (here is [construction](https://github.com/AlexeyMerzlyakov/navigation2/blob/52d2c12628f3c3d2ef8225fe97b0a9fe9805a5db/nav2_map_server/src/vo_server/vector_object_server.cpp#L390) of new Polygon).
Inside both Polygon and Circle constructors, if UUID was specified by developer in service request, we just passing it to shape' parameters, as `params_ = params;`. If not set, [generating](https://github.com/AlexeyMerzlyakov/navigation2/blob/52d2c12628f3c3d2ef8225fe97b0a9fe9805a5db/nav2_map_server/src/vo_server/vector_object_shapes.cpp#L100) a new UUID (which could be obtained later by calling `get_shapes` service).

During the update of existing Polygon/Circle, UUID is being [searched](https://github.com/AlexeyMerzlyakov/navigation2/blob/52d2c12628f3c3d2ef8225fe97b0a9fe9805a5db/nav2_map_server/src/vo_server/vector_object_server.cpp#L355) (I'll try to replace it with unordered map, in accordance with comment from above) by requested by service call UUID.

Reply:
This seems a little convoluted. to add new virtual parameters for incoming shapes rather than storing the attributes of the shape within the shape itself (and params only gotten for the shapes set in files on initialization -- then no more ROS parameters in the system).

Reply:
Hm... It already works like you've mentioned: these parameters are not virtual. Abstract class `Shape` does not contain any of shape parameters. Parameters are already belonging to each shape. For `Polygon` shape class it is:
```
class Polygon : public Shape
{
...
protected:
  nav2_msgs::msg::PolygonVO::SharedPtr params_;
};
```
And for `Circle` shape class it is:
```
class Circle : public Shape
{
...
protected;
  nav2_msgs::msg::CircleVO::SharedPtr params_;
};
```

> and params only gotten for the shapes set in files on initialization -- then no more ROS parameters in the system

This is also how it already works: ROS-parameters are being set during initialization stage inside `Polygon::getROSParameters()` and `Circle::getROSParameters()`. Aha, these methods could be called from polygon/circle constructor. I've got it, if this is what you mean in this review comment.

Reply:
Afte code double-checking, `Polygon`/`Circle` parameters initialization / setting were simplified:
Now the `Polygon`/`Circle` constructor is almost empty. Two routines are responsible for shape/parameters setting:

- `obtainParams(name)` - used to obtain ROS-parameters. Called once during ROS-node configuring.
- `setParams(params)` - used to set ROS-parameters during add/update shape service call.

The changes were reflected in https://github.com/ros-planning/navigation2/pull/3930/commits/f1fe2479025920c4927c4931ea9fd9f01910a93b